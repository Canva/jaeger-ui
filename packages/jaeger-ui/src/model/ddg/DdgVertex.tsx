// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TVertex } from '@jaegertracing/plexus/lib/types';

import { DdgEdge, PathElem } from './types';

type TDdgVertex = {
  egressEdges: Map<DdgVertex, DdgEdge>;
  ingressEdges: Map<DdgVertex, DdgEdge>;
  pathElems: Set<PathElem>;
};

export default class DdgVertex implements TVertex<TDdgVertex> {
  egressEdges: Map<DdgVertex, DdgEdge>;
  key: string;
  ingressEdges: Map<DdgVertex, DdgEdge>;
  pathElems: Set<PathElem>;

  constructor({ key }: { key: string }) {
    this.egressEdges = new Map();
    this.key = key;
    this.ingressEdges = new Map();
    this.pathElems = new Set();
  }

  toJSON() {
    type TDigestibleEdges = Record<string, DdgEdge>;
    const digestibleEgressEdges: TDigestibleEdges = {};
    const digestibleIngressEdges: TDigestibleEdges = {};
    this.egressEdges.forEach((edge, egreesNeighbor) => {
      digestibleEgressEdges[egreesNeighbor.key] = edge;
    });
    this.ingressEdges.forEach((edge, ingreesNeighbor) => {
      digestibleIngressEdges[ingreesNeighbor.key] = edge;
    });
    return {
      egressEdges: digestibleEgressEdges,
      key: this.key,
      ingressEdges: digestibleIngressEdges,
      pathElems: this.pathElems,
    };
  }

  toString() {
    return JSON.stringify(this.toJSON(), null, 2);
  }

  get [Symbol.toStringTag]() {
    return `DdgVertex ${this.key}`;
  }
}
