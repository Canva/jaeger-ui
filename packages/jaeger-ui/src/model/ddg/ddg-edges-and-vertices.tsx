// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import _map from 'lodash/map';

import { compareVisibilityKeys, changeVisibility } from './visibility-key';

import { PathElem, TDdgEdge, TDdgEdgeKeys, /* TDdgVertex, */ DdgVertex, TDdgModel, TDdgPathElemsByDistance, TDdgPath, TDdgServiceMap } from './types';

// TODO rename file
export default class DdgEdgesAndVertices {
  lastVisibilityKey: string;
  pathElemsByDistance: TDdgPathElemsByDistance;
  pathElemToVertex: Map<PathElem, DdgVertex>;
  paths: TDdgPath[];
  services: TDdgServiceMap;
  // edges: TDdgEdge[];
  edges: Set<TDdgEdge>;
  vertices: Map<string, DdgVertex>;
  visibilityIdxToPathElem: Map<number, PathElem>;

  // flow all wrong
  constructor({ ddgModel, visibilityKey }: { ddgModel: TDdgModel, visibilityKey: string }) {
    this.pathElemsByDistance = ddgModel.pathElemsByDistance;
    this.paths = ddgModel.paths;
    this.services = ddgModel.services;
    this.visibilityIdxToPathElem = ddgModel.visibilityIdxToPathElem;
    this.pathElemToVertex = new Map();
    this.edges = new Set();
    this.vertices = new Map();
    this.lastVisibilityKey = visibilityKey;

    const visibleIndices = compareVisibilityKeys({ oldVisibilityKey: '', newVisibilityKey: visibilityKey }).added;

    this.showPathElems(visibleIndices);
  }

  private showPathElems = (newIndices: number[]) => {
    newIndices.forEach(newIdx => {
      const pathElem = this.visibilityIdxToPathElem.get(newIdx);
      if (!pathElem) {
        throw new Error(`Given visibilityIdx: "${newIdx}" that does not exist`);
      }
      const key = this.getVertexKey(pathElem);

      let vertex = this.vertices.get(key);
      if (!vertex) {
        vertex = new DdgVertex({ key });
        this.vertices.set(key, vertex);
      }

      this.pathElemToVertex.set(pathElem, vertex);
      vertex.pathElems.add(pathElem);

      const connectedPathElem = pathElem.focalSideNeighbor;
      if (connectedPathElem) {
        const connectedVertex = this.pathElemToVertex.get(connectedPathElem);
        if (!connectedVertex) {
          // TODO: Improve error message
          throw new Error(`Non-focal pathElem lacks precursorVertex. PathElem: ${JSON.stringify(pathElem, null, 2)}`);
        }

        if (!vertex[pathElem.focalSideEdgesKey].has(connectedVertex)) {
          const newEdge: TDdgEdge = pathElem.focalSideEdgesKey === 'ingressEdges'
            ? {
              from: connectedVertex,
              to: vertex,
            }
            : {
              from: vertex,
              to: connectedVertex,
            };
          vertex[pathElem.focalSideEdgesKey].set(connectedVertex, newEdge);
          connectedVertex[pathElem.farSideEdgesKey].set(vertex, newEdge);
          this.edges.add(newEdge);
        }
      }
    });
  }

  // This function assumes the density is set to PPE with distinct operations
  // class property so that it can be aware of density in late-alpha
  //
  // might make sense to live on PathElem so that pathElems can be compared when checking how many
  // inbound/outbound edges are visible for a vertex
  private getVertexKey = (pathElem: PathElem): string => {
    const { distance, memberOf } = pathElem;
    const { focalIdx, members } = memberOf;
    const startIdx = Math.min(focalIdx, focalIdx + distance);

    return members.slice(startIdx, startIdx + Math.abs(distance) + 1)
      .map(({ operation }) => `${operation.service.name}::${operation.name}`).join('|');
  }
}
