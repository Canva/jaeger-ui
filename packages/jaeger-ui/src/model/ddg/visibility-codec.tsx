// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* eslint-disable no-bitwise */

import memoize from 'lru-memoize';

// This bucket size was chosen as JavaScript uses 32-bit numbers for bitwise operators, so the maximum number
// of indices that can be tracked in a single number is 32. Increasing from 31 visibility values per base36
// number to 32 visibility values would require an additional base36 character, which reduces the efficiency.
// For more info: https://repl.it/repls/IllfatedMaroonObservation
const VISIBILITY_BUCKET_SIZE = 31;

/*
 * Converts string csv of base36 numbers into array of visible indices.
 *
 * @param {string} encoded - base36 csv visibility encoding.
 * @returns {number[]} - Visibile indices for this encoding.
 */
export const decode: (encoded: string) => number[] = memoize(10)((encoded: string): number[] => {
  const rv: number[] = [];
  encoded.split(',').forEach((partial, i) => {
    const partialAsNumber = partial ? parseInt(partial, 36) : 0;
    for (let j = 0; partialAsNumber >= 1 << j; j += 1) {
      if ((1 << j) & partialAsNumber) {
        rv.push(i * VISIBILITY_BUCKET_SIZE + j);
      }
    }
  });
  return rv;
});

/*
 * Given a visibility index, returns the index of the bucket that contains the given index and the power of
 * two that would reflect the given index being visible, relative to the bucket it resides in.
 *
 * @param {number} visibilityIdx - An index within the visibility key
 * @returns {object} - Two numbers relating to the given number.
 *     bucketIdx - Index of the JavaScript number in getBuckets that contains the visibility data for the
 *         given index.
 *     visibilityValue - A power of two that can be used with getBuckets(key)[bucketIdx] to determine if the
 *         given index is visible, or to change its visibility.
 */
function convertAbsoluteIdxToRelativeValues(absIdx: number) {
  const csvIdx = Math.floor(absIdx / VISIBILITY_BUCKET_SIZE);
  const relativeIdx = absIdx % VISIBILITY_BUCKET_SIZE;
  const visibilityValue = 1 << relativeIdx;
  return { csvIdx, visibilityValue };
}

/*
 * Converts array of visible indices into string csv of base36 numbers.
 *
 * @param {number[]} decoded - Visibile indices for this encoding.
 * @returns {string} - base36 csv visibility encoding.
 */
export const encode = (decoded: number[]): string => {
  const partial: number[] = [];
  decoded.forEach(visIdx => {
    const { csvIdx, visibilityValue } = convertAbsoluteIdxToRelativeValues(visIdx);
    partial[csvIdx] |= visibilityValue;
  });
  return partial.map(p => p.toString(36)).join();
};
