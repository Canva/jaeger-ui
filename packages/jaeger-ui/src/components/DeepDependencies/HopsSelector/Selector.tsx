// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Popover } from 'antd';
import cx from 'classnames';
import SortAmountAsc from 'react-icons/lib/fa/sort-amount-asc.js';
import IoChevronDown from 'react-icons/lib/io/chevron-down';
import IoChevronRight from 'react-icons/lib/io/chevron-right';

import { ECheckedStatus, EDirection, THop } from '../../../model/ddg/types';

import './Selector.css';

type TProps = {
  direction: EDirection;
  furthestDistance: number;
  furthestFullness: ECheckedStatus;
  handleClick: (distance: number, direction: EDirection) => void;
  hops: THop[];
};

const CLASSNAME = 'HopsSelector--Selector';

export default class Selector extends PureComponent<TProps> {
  private makeBtn = ({ distance, fullness, suffix }: THop & { suffix?: string }) => {
    const { handleClick, direction } = this.props;
    return (
      <button
        key={`${distance} ${direction} ${suffix}`}
        className={cx(`${CLASSNAME}--btn`, fullness, suffix || 'popover-content')}
        type="button"
        onClick={() => handleClick(distance, direction)}
      >
        {!suffix && <IoChevronRight className={`${CLASSNAME}--ChevronRight`} />} {Math.abs(distance)}
      </button>
    );
  };

  render() {
    const { direction, furthestDistance, furthestFullness, handleClick, hops } = this.props;
    const streamText = direction === EDirection.Downstream ? 'Down' : 'Up';
    const streamLabel = `${streamText}stream hops`;
    const lowercaseLabel = streamLabel.toLowerCase();
    if (hops.length === 1) {
      return <span className={CLASSNAME}>No {lowercaseLabel}</span>;
    }

    const decrementBtn = (
      <button
        key={`decrement ${direction}`}
        disabled={furthestDistance === 0}
        className={cx(CLASSNAME, 'decrement')}
        type="button"
        onClick={() => handleClick(furthestDistance - direction, direction)}
      >
        -
      </button>
    );
    const incrementBtn = (
      <button
        key={`increment ${direction}`}
        disabled={furthestDistance === hops[hops.length - 1].distance}
        className={cx(CLASSNAME, 'increment')}
        type="button"
        onClick={() => handleClick(furthestDistance + direction, direction)}
      >
        +
      </button>
    );
    const delimiterBtn = this.makeBtn({
      ...hops[hops.length - 1],
      suffix: 'delimiter',
    });
    const furthestBtn = this.makeBtn({
      distance: furthestDistance,
      fullness: furthestFullness,
      suffix: 'furthest',
    });

    return (
      <Popover
        arrowPointAtCenter
        content={[decrementBtn, ...hops.map(this.makeBtn), incrementBtn]}
        placement="bottom"
        visible={direction === EDirection.Upstream}
        title={`Visible ${lowercaseLabel}`}
      >
        <span className={CLASSNAME}>
          <SortAmountAsc className={cx(`${CLASSNAME}--AscIcon`, streamText)} /> {streamLabel} {furthestBtn} /{' '}
          {delimiterBtn} <IoChevronDown className={`${CLASSNAME}--ChevronDown`} />
        </span>
      </Popover>
    );
  }
}
