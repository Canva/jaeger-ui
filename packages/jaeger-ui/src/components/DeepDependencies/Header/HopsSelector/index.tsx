// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';

import { decode, encodeDistance } from '../../../../model/ddg/visibility-codec';
import { ECheckedStatus, EDirection, TDdgModel, THop } from '../../../../model/ddg/types';
import Selector from './Selector';

type TProps = {
  ddgModel: TDdgModel;
  updateVisEncoding: (visEncoding: string) => void;
  visEncoding?: string;
};

export default class HopsSelector extends PureComponent<TProps> {
  private handleClick = (distance: number, direction: EDirection) => {
    const { ddgModel, updateVisEncoding, visEncoding } = this.props;

    updateVisEncoding(
      encodeDistance({
        ddgModel,
        direction,
        distance,
        prevVisEncoding: visEncoding,
      })
    );
  };

  render() {
    const { ddgModel, visEncoding } = this.props;
    const { distanceToPathElems } = ddgModel;

    const downstreamHops: THop[] = [];
    const upstreamHops: THop[] = [];
    let minVisDistance = 0;
    let minVisDistanceFullness = ECheckedStatus.Empty;
    let maxVisDistance = 0;
    let maxVisDistanceFullness = ECheckedStatus.Empty;
    const visibleIndices = visEncoding && new Set(decode(visEncoding));

    distanceToPathElems.forEach((elems, distance) => {
      let fullness: ECheckedStatus;
      if (visibleIndices) {
        const visible = elems.filter(({ visibilityIdx }) => visibleIndices.has(visibilityIdx));
        if (visible.length === elems.length) {
          fullness = ECheckedStatus.Full;
        } else if (!visible.length) {
          fullness = ECheckedStatus.Empty;
        } else {
          fullness = ECheckedStatus.Partial;
        }
      } else {
        fullness = Math.abs(distance) <= 2 ? ECheckedStatus.Full : ECheckedStatus.Empty;
      }

      if (distance >= 0) downstreamHops.push({ distance, fullness });
      if (distance <= 0) upstreamHops.push({ distance, fullness });

      if (fullness !== ECheckedStatus.Empty) {
        if (distance >= maxVisDistance) {
          maxVisDistance = distance;
          maxVisDistanceFullness = fullness;
        }
        if (distance <= minVisDistance) {
          minVisDistance = distance;
          minVisDistanceFullness = fullness;
        }
      }
    });

    downstreamHops.sort(({ distance: a }, { distance: b }) => a - b);
    upstreamHops.sort(({ distance: a }, { distance: b }) => b - a);

    return (
      <div>
        <Selector
          direction={EDirection.Upstream}
          handleClick={this.handleClick}
          hops={upstreamHops}
          furthestDistance={minVisDistance}
          furthestFullness={minVisDistanceFullness}
        />
        <Selector
          direction={EDirection.Downstream}
          handleClick={this.handleClick}
          hops={downstreamHops}
          furthestDistance={maxVisDistance}
          furthestFullness={maxVisDistanceFullness}
        />
      </div>
    );
  }
}
