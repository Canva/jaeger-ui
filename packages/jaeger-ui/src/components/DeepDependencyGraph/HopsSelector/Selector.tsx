// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { Popover } from 'antd';
import cx from 'classnames';

import { EStream, EFullness, THop } from './types';

import './Selector.css';

type TProps = {
  furthestDistance: number;
  furthestFullness: EFullness;
  handleClick: (distance: number, direction: EStream) => void;
  hops: THop[];
  stream: EStream;
};

export default class Selector extends PureComponent<TProps> {
  private btn = ({ distance, fullness, suffix = 'popover-content' }: THop & { suffix?: string }) => {
    const { handleClick, stream } = this.props;
    return (
      <button
        key={`${distance} ${stream} ${suffix}`}
        className={cx('Selector--btn', fullness, suffix)}
        type="button"
        onClick={() => handleClick(distance, stream)}
      >
        {Math.abs(distance)}
      </button>
    );
  };

  render() {
    const { furthestDistance, furthestFullness, handleClick, hops, stream } = this.props;
    const streamLabel = `${stream === EStream.down ? 'Down' : 'Up'}stream hops`;
    const lowercaseLabel = streamLabel.toLowerCase();
    if (hops.length === 1) {
      return <span className="Selector">No {lowercaseLabel}</span>;
    }

    const decrementBtn = (
      <button
        key={`decrement ${stream}`}
        disabled={furthestDistance === 0}
        type="button"
        onClick={() => handleClick(furthestDistance - stream, stream)}
      >
        -
      </button>
    );
    const incrementBtn = (
      <button
        key={`increment ${stream}`}
        disabled={furthestDistance === hops[hops.length - 1].distance}
        type="button"
        onClick={() => handleClick(furthestDistance + stream, stream)}
      >
        +
      </button>
    );
    const delimiterBtn = this.btn({
      ...hops[hops.length - 1],
      suffix: 'delimiter',
    });
    const furthestBtn = this.btn({
      distance: furthestDistance,
      fullness: furthestFullness,
      suffix: 'furthest',
    });

    return (
      <Popover
        arrowPointAtCenter
        content={[decrementBtn, hops.map(this.btn), incrementBtn]}
        placement="bottom"
        title={`Visible ${lowercaseLabel}`}
      >
        <span className="Selector">
          {streamLabel} {furthestBtn} / {delimiterBtn}
        </span>
      </Popover>
    );
  }
}
