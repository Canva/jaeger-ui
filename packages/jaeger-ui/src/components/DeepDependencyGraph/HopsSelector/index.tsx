// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';

import { decode, encode } from '../../../model/ddg/visibility-codec';
import { TDdgModel } from '../../../model/ddg/types';
import Selector from './Selector';

import { EStream, EFullness, THop } from './types';

type TProps = {
  ddgModel: TDdgModel;
  updateVisEncoding: (visEncoding: string) => void;
  visEncoding?: string;
};

export default class HopsSelector extends PureComponent<TProps> {
  private handleClick = (clickedDistance: number, direction: EStream) => {
    const { ddgModel, updateVisEncoding, visEncoding } = this.props;
    const { distanceToPathElems, visIdxToPathElem } = ddgModel;

    let nextVisible: number[];
    if (visEncoding) {
      nextVisible = decode(visEncoding).filter(
        idx => visIdxToPathElem[idx] && Math.sign(visIdxToPathElem[idx].distance) !== direction
      );
    } else {
      nextVisible = [
        ...(distanceToPathElems.get(-1 * direction) || []),
        ...(distanceToPathElems.get(-2 * direction) || []),
      ].map(({ visibilityIdx }) => visibilityIdx);
    }

    for (let distance = 0; distance !== clickedDistance + direction; distance += direction) {
      const elems = distanceToPathElems.get(distance);
      if (elems) {
        nextVisible.push(...elems.map(({ visibilityIdx }) => visibilityIdx));
      }
    }

    updateVisEncoding(encode(nextVisible));
  };

  render() {
    const { ddgModel, visEncoding } = this.props;
    const { distanceToPathElems } = ddgModel;

    const hops: THop[] = [];
    let minVisDistance = 0;
    let minVisDistanceFullness = EFullness.empty;
    let maxVisDistance = 0;
    let maxVisDistanceFullness = EFullness.empty;
    const visibleIndices = visEncoding && new Set(decode(visEncoding));

    distanceToPathElems.forEach((elems, distance) => {
      let fullness: EFullness;
      if (visibleIndices) {
        const visible = elems.filter(({ visibilityIdx }) => visibleIndices.has(visibilityIdx));
        if (visible.length === elems.length) {
          fullness = EFullness.full;
        } else if (!visible.length) {
          fullness = EFullness.empty;
        } else {
          fullness = EFullness.partial;
        }
      } else {
        fullness = Math.abs(distance) <= 2 ? EFullness.full : EFullness.empty;
      }

      hops.push({ distance, fullness });
      if (fullness !== EFullness.empty) {
        if (distance > maxVisDistance) {
          maxVisDistance = distance;
          maxVisDistanceFullness = fullness;
        } else if (distance < minVisDistance) {
          minVisDistance = distance;
          minVisDistanceFullness = fullness;
        }
      }
    });

    hops.sort(({ distance: a }, { distance: b }) => a - b);
    const splitIdx = hops.findIndex(({ distance }) => !distance);
    const upstream = hops.slice(0, splitIdx + 1).reverse();
    const downstream = hops.slice(splitIdx);

    return (
      <div>
        <Selector
          handleClick={this.handleClick}
          hops={upstream}
          furthestDistance={minVisDistance}
          furthestFullness={minVisDistanceFullness}
          stream={EStream.up}
        />
        <Selector
          handleClick={this.handleClick}
          hops={downstream}
          furthestDistance={maxVisDistance}
          furthestFullness={maxVisDistanceFullness}
          stream={EStream.down}
        />
      </div>
    );
  }
}
