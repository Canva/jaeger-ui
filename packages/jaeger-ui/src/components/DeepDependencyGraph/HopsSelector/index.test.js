// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';

import {
  focalPayloadElem,
  longSimplePath,
  shortPath,
  simplePath,
} from '../../../model/ddg/sample-paths.test.resources';
import transformDdgData from '../../../model/ddg/transformDdgData';
import * as codec from '../../../model/ddg/visibility-codec';
import HopsSelector from '.';

describe('HopsSelector', () => {
  const ddgModel = transformDdgData([longSimplePath, simplePath], focalPayloadElem);
  const shortModel = transformDdgData([shortPath], focalPayloadElem);

  describe('handleClick', () => {
    const updateVisEncoding = () => {};
    let encodeSpy;

    beforeAll(() => {
      encodeSpy = jest.spyOn(codec, 'encode');
    });

    function validateEncoding(start, end, omit) {
      const expectedIndices = [];
      for (let i = start; i <= end; i++) {
        expectedIndices.push(
          ...ddgModel.distanceToPathElems
            .get(i)
            .map(({ visibilityIdx }) => visibilityIdx)
            .filter(idx => !omit || !omit.has(idx))
        );
      }
      const { calls } = encodeSpy.mock;
      expect(new Set(calls[calls.length - 1][0])).toEqual(new Set(expectedIndices));
    }

    describe('without visEncoding', () => {
      const hopsSelector = new HopsSelector({ ddgModel, updateVisEncoding });

      it('removes hops when selecting less than 2 hops', () => {
        hopsSelector.handleClick(-1, -1);
        validateEncoding(-1, 2);
      });

      it('adds hops when selecting more than 2 hops', () => {
        hopsSelector.handleClick(4, 1);
        validateEncoding(-2, 4);
      });

      it('handles selecting 0', () => {
        hopsSelector.handleClick(0, 1);
        validateEncoding(-2, 0);

        hopsSelector.handleClick(0, -1);
        validateEncoding(0, 2);
      });

      it('handles DDGs smaller than two hops', () => {
        const shortSelector = new HopsSelector({ ddgModel: shortModel, updateVisEncoding });
        shortSelector.handleClick(0, -1);
        const { calls } = encodeSpy.mock;
        expect(calls[calls.length - 1][0]).toEqual([0]);
      });

      it('handles out of bounds selection', () => {
        hopsSelector.handleClick(8, 1);
        validateEncoding(-2, 6);
      });
    });

    describe('with visEncoding', () => {
      const fourHops = codec.encode([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]);
      const fourHopsSelector = new HopsSelector({ ddgModel, updateVisEncoding, visEncoding: fourHops });
      const partialHops = codec.encode([0, 1, 2, 3, 4, 5, 7, 9, 10, 11]);
      const partialHopsSelector = new HopsSelector({ ddgModel, updateVisEncoding, visEncoding: partialHops });

      it('adds hops when selecting more hops', () => {
        fourHopsSelector.handleClick(6, 1);
        validateEncoding(-4, 6);
      });

      it('removes hops when selecting fewer hops', () => {
        fourHopsSelector.handleClick(-2, -1);
        validateEncoding(-2, 4);
      });

      it('handles partially full hops', () => {
        partialHopsSelector.handleClick(4, 1);
        validateEncoding(-3, 4, new Set([8]));
      });

      it('handles selecting 0', () => {
        partialHopsSelector.handleClick(0, 1);
        validateEncoding(-3, 0, new Set([8]));

        partialHopsSelector.handleClick(0, -1);
        validateEncoding(0, 3, new Set([6]));
      });
    });
  });

  describe('render', () => {
    describe('without visEncoding', () => {
      it('renders hops within two hops as full and others as empty', () => {
        const wrapper = shallow(<HopsSelector ddgModel={ddgModel} />);
        expect(wrapper).toMatchSnapshot();
      });

      it('handles DDGs smaller than two hops', () => {
        const wrapper = shallow(<HopsSelector ddgModel={shortModel} />);
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('with visEncoding', () => {
      it('renders hops with correct fullness', () => {
        const visEncoding = codec.encode([0, 1, 2, 3, 4, 5, 7, 9, 10, 11, 12, 13]);
        const wrapper = shallow(<HopsSelector ddgModel={ddgModel} visEncoding={visEncoding} />);
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
