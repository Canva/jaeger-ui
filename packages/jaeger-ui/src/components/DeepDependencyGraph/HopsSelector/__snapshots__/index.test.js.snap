// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HopsSelector render with visEncoding renders hops with correct fullness 1`] = `
<div>
  <Selector
    furthestDistance={-4}
    furthestFullness="full"
    handleClick={[Function]}
    hops={
      Array [
        Object {
          "distance": 0,
          "fullness": "full",
        },
        Object {
          "distance": -1,
          "fullness": "full",
        },
        Object {
          "distance": -2,
          "fullness": "partial",
        },
        Object {
          "distance": -3,
          "fullness": "full",
        },
        Object {
          "distance": -4,
          "fullness": "full",
        },
        Object {
          "distance": -5,
          "fullness": "empty",
        },
        Object {
          "distance": -6,
          "fullness": "empty",
        },
      ]
    }
    stream={-1}
  />
  <Selector
    furthestDistance={4}
    furthestFullness="full"
    handleClick={[Function]}
    hops={
      Array [
        Object {
          "distance": 0,
          "fullness": "full",
        },
        Object {
          "distance": 1,
          "fullness": "full",
        },
        Object {
          "distance": 2,
          "fullness": "partial",
        },
        Object {
          "distance": 3,
          "fullness": "full",
        },
        Object {
          "distance": 4,
          "fullness": "full",
        },
        Object {
          "distance": 5,
          "fullness": "empty",
        },
        Object {
          "distance": 6,
          "fullness": "empty",
        },
      ]
    }
    stream={1}
  />
</div>
`;

exports[`HopsSelector render without visEncoding handles DDGs smaller than two hops 1`] = `
<div>
  <Selector
    furthestDistance={-1}
    furthestFullness="full"
    handleClick={[Function]}
    hops={
      Array [
        Object {
          "distance": 0,
          "fullness": "full",
        },
        Object {
          "distance": -1,
          "fullness": "full",
        },
      ]
    }
    stream={-1}
  />
  <Selector
    furthestDistance={0}
    furthestFullness="empty"
    handleClick={[Function]}
    hops={
      Array [
        Object {
          "distance": 0,
          "fullness": "full",
        },
      ]
    }
    stream={1}
  />
</div>
`;

exports[`HopsSelector render without visEncoding renders hops within two hops as full and others as empty 1`] = `
<div>
  <Selector
    furthestDistance={-2}
    furthestFullness="full"
    handleClick={[Function]}
    hops={
      Array [
        Object {
          "distance": 0,
          "fullness": "full",
        },
        Object {
          "distance": -1,
          "fullness": "full",
        },
        Object {
          "distance": -2,
          "fullness": "full",
        },
        Object {
          "distance": -3,
          "fullness": "empty",
        },
        Object {
          "distance": -4,
          "fullness": "empty",
        },
        Object {
          "distance": -5,
          "fullness": "empty",
        },
        Object {
          "distance": -6,
          "fullness": "empty",
        },
      ]
    }
    stream={-1}
  />
  <Selector
    furthestDistance={2}
    furthestFullness="full"
    handleClick={[Function]}
    hops={
      Array [
        Object {
          "distance": 0,
          "fullness": "full",
        },
        Object {
          "distance": 1,
          "fullness": "full",
        },
        Object {
          "distance": 2,
          "fullness": "full",
        },
        Object {
          "distance": 3,
          "fullness": "empty",
        },
        Object {
          "distance": 4,
          "fullness": "empty",
        },
        Object {
          "distance": 5,
          "fullness": "empty",
        },
        Object {
          "distance": 6,
          "fullness": "empty",
        },
      ]
    }
    stream={1}
  />
</div>
`;
